src/csharp/Grpc.Core/Internal/CallSafeHandle.cs

```diff
@@ -127,16 +128,52 @@ public void StartSendCloseFromClient(ISendCompletionCallback callback)
             }
         }
 
-        public void StartSendStatusFromServer(ISendStatusFromServerCompletionCallback callback, Status status, MetadataArraySafeHandle metadataArray, bool sendEmptyInitialMetadata,
+        public unsafe void StartSendStatusFromServer(ISendStatusFromServerCompletionCallback callback, Status status, MetadataArraySafeHandle metadataArray, bool sendEmptyInitialMetadata,
             byte[] optionalPayload, WriteFlags writeFlags)
         {
             using (completionQueue.NewScope())
             {
                 var ctx = completionQueue.CompletionRegistry.RegisterBatchCompletion(CompletionHandler_ISendStatusFromServerCompletionCallback, callback);
                 var optionalPayloadLength = optionalPayload != null ? new UIntPtr((ulong)optionalPayload.Length) : UIntPtr.Zero;
-                var statusDetailBytes = MarshalUtils.GetBytesUTF8(status.Detail);
-                Native.grpcsharp_call_send_status_from_server(this, ctx, status.StatusCode, statusDetailBytes, new UIntPtr((ulong)statusDetailBytes.Length), metadataArray, sendEmptyInitialMetadata ? 1 : 0,
-                    optionalPayload, optionalPayloadLength, writeFlags).CheckOk();
+
+                const int MaxStackAllocBytes = 256;
+                int maxBytes = MarshalUtils.GetMaxByteCountUTF8(status.Detail);
+                if (maxBytes > MaxStackAllocBytes)
+                {
+                    // pay the extra to get the *actual* size; this could mean that
+                    // it ends up fitting on the stack after all, but even if not
+                    // it will mean that we ask for a *much* smaller buffer
+                    maxBytes = MarshalUtils.GetByteCountUTF8(status.Detail);
+                }
+
+                if (maxBytes <= MaxStackAllocBytes)
+                {   // for small status, we can encode on the stack without touching arrays
+                    // note: if init-locals is disabled, it would be more efficient
+                    // to just stackalloc[MaxStackAllocBytes]; but by default, since we
+                    // expect this to be small and it needs to wipe, just use maxBytes
+                    byte* ptr = stackalloc byte[maxBytes];
+                    int statusBytes = MarshalUtils.GetBytesUTF8(status.Detail, ptr, maxBytes);
+                    Native.grpcsharp_call_send_status_from_server(this, ctx, status.StatusCode, ptr, new UIntPtr((ulong)statusBytes), metadataArray, sendEmptyInitialMetadata ? 1 : 0,
```

Comment:
just FYI, even with this optimization, the data gets copied in the native code because it needs to be converted to grpc_slice that needs to stay valid until the send_status_from_server operation finishes.

https://github.com/grpc/grpc/blob/d7bd178b31fb4953db17c72a238bd745cb4d9ca6/src/csharp/ext/grpc_csharp_ext.c#L800

Your change avoids at least one of the copies, so there's an improvement, I just wanted to make it clear that there are still opportunities to improve further.

Reply:
indeed; with some future work we would expose an API that is basically `IntPtr RequestSliceFromNative(int length)` (assuming we're talking single-slice), and just write direct to the slice; non-trivial changes, and needs a refactor to `grpcsharp_call_send_status_from_server` obviously, but... intriguing; it basically defers the entire "stackalloc vs rent vs allocate" problem down the the native lib, which is *already* doing that

Reply:
thought: the right time to consider this would probably be when revisiting `IBufferWriter` for `optionalPayload` - then *everything* gets written direct into the unmanaged/native buffers